openssh: add {Allow,Deny}{Users,Groups} settings

christoph-heiss opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

From openssh's perspective, the `AllowUsers`, `DenyUsers`, `AllowGroups` and `DenyGroups` settings are space-separated lists of user-/groupnames, thus treat them the same here.

This allows them to be set like `services.openssh.settings.AllowUsers = [ "foo" "bar" ]` and be merged automatically. 

~~[ Not sure if this is considered a breaking change, since the freefrom `services.openssh.settings` option is not yet part of any stable release. If it is, please let me know and I add a release notes entry as appropriate! ]~~

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

christoph-heiss committed:
openssh: add {Allow,Deny}{Users,Groups} settings

`settingsFormat` unfortunaly needed a bit of a rework, see also
discussion in #227442.

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

teto commented:
one of the advantage of `settings` IMO is that you dont have to explicit the options, thus helping with maintainance.
You had to explicit those because of mkValueStringSshd concatenating with `,` ?
I was exploring an alternative in https://github.com/NixOS/nixpkgs/pull/215397 with ` listsAsDuplicateKeys = true;` but I was worried it would result in inconsistent behavior for some options 
https://github.com/NixOS/nixpkgs/pull/215397/files#diff-909cff3303f72af4beab9cc703252041321d4df4d5fe94729505a35165787fa4R25

To answer your last questions, your suggestion is not a breaking change.

christoph-heiss commented:
Yes, exactly. I've ended up with `,`-concatenated usernames, which .. well, doesn't work as you can imagine. So having these settings be (correctly) mergeable would be benefical and what users would expect, anyway, IHMO.

`listsAsDuplicateKeys = true` should work well for these keys, according to the `sshd_config(5)` man page ("This keyword may appear multiple times in sshd_config with each instance appending to the list."). Didn't know of this option until now.

If that would be the preferred way, I will take a look at it. At least in this case there should be no inconsistencies - as these directives are processed in a fixed order anyway ("The allow/deny users directives are processed in the following order: DenyUsers, AllowUsers.", from the man page again).

Thanks for having a look (and answering my question)!

teto commented:
> Edit: Reading the openssh source code, for the above four directives the value indeed gets appended if it appears multiple time ...  but not for e.g. the Macs directive ([see here](https://github.com/openssh/openssh-portable/blob/ac383f3a5c6f529a2e8a5bc44af79a08c7da294e/servconf.c#L1902)), where it simply skips any further values.

Yeah openssh is inconsistent with this, which is quite annoying. Enabling `listsAsDuplicateKeys = true`  means that scattering different `settings.Port` across files, ssh would now listen on all of them instead of erroring out because of different values (I think). Which is why I had stopped investigating :(

christoph-heiss commented:
Finally came around to it and fixed the eval failure. I've also added a testcase for `AllowUsers`, to make it easier to verify its functionality. (The other three options are pretty much identical, so one should be enough.)

christoph-heiss commented:
Ping. Is this still something that could get merged?
I'm happy to improve on it further, whatever it takes to get it merged.

tomfitzhenry reviewed (approved):
+1 to this being useful: the default merge of concat-with-commas doesn't work well with the space-delimited options such as `AllowUsers`. Merge semantics work well with Allow/Deny in SSH, since the entries on each directive are order-independent.

LGTM:
* Code looks good: works, is minimal, well documented, links to manpages work. 
* Test looks good, and passes.

aanderse commented:
I'll take a look at this PR in the next few days.

christoph-heiss committed:
nixos/tests/openssh: add test for `AllowUsers`

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

aanderse commented:
@christoph-heiss instead of using `apply`, which as far as i can recall has some problems associated with it the way you are using it - correct me if i'm wrong please, what do you think about taking care of this file format issue in a more general way, like so?

```
  settingsFormat = { ... }:
  let
    # reports boolean as yes / no
    mkValueStringSshd = with lib; v:
      if isInt           v then toString v
      else if isString   v then v
      else if true  ==   v then "yes"
      else if false ==   v then "no"
      else throw "unsupported type ${builtins.typeOf v}: ${(lib.generators.toPretty {}) v}";

    base = pkgs.formats.keyValue {
      listsAsDuplicateKeys = true;
      mkKeyValue = lib.generators.mkKeyValueDefault { mkValueString = mkValueStringSshd; } " ";
    };
  in
  {
    type = base.type;

    generate = name: value:
      let
        transformedValue = mapAttrs (key: val:
          if isList val then
            if
              key == "Ciphers" ||
              key == "KexAlgorithms" ||
              key == "Macs"
            then
              concatStringsSep "," val
            else if
              key == "AllowGroups" ||
              key == "AllowUsers" ||
              key == "DenyGroups" ||
              key == "DenyUsers" ||
              key == "AuthorizedKeysFile"
            then
              concatStringsSep " " val
            else
              val
          else
            val
        ) (filterAttrs (_: v: v != null) value);
      in base.generate name transformedValue;
  };
```

the above encapsulates all of the quirky behaviour involved in coercing `sshd_config` to render properly in one spot instead of speckled throughout the module

thoughts?

cc @infinisil because of https://github.com/NixOS/nixpkgs/pull/215397#issuecomment-1751656301

christoph-heiss commented:
@aanderse Removing quirky behavior and having to look up in the opensssh source whether values are appended or replaced when the option is present multiple times is definitely a good thing, I guess.

I'd be happy to implement the approach you drafted - if @teto would also be OK with that, before I change/implement something.

aanderse reviewed (approved):
awesome! thanks for doing this

let's pause this until branch off of 23.11 and then hit merge - sound good to everyone?

teto commented:
> let's pause this until branch off of 23.11 and then hit merge - sound good to everyone?

why the wait ? I think there is enough time before branch off to spot issues .

aanderse commented:
ok if you're feeling bold let's do it then! 😉

everyone ready to click "merge"?

infinisil commented on nixos/modules/services/networking/ssh/sshd.nix:
```diff
@@ -12,22 +12,38 @@ let
+      base = pkgs.formats.keyValue {
+        mkKeyValue = lib.generators.mkKeyValueDefault { inherit mkValueString; } " ";
+      };
+      commaSeparated = [ "Ciphers" "KexAlgorithms" "Macs" ];
+      spaceSeparated = [ "AuthorizedKeysFile" "AllowGroups" "AllowUsers" "DenyGroups" "DenyUsers" ];
```
Can you add a comment here explaining why these are special-cased, whether more are needed in the future, and how this can be determined?

christoph-heiss replied:
I have added a comment trying to explain the situation with these options and where one can find out the right case for them, if more need to be added in the future.

teto commented:
If the maintainer @aanderse is fine with merging this, I  would let @christoph-heiss decide on timing: I can relate to a contributor wanting to merge something in time for the release, but at the same time it can be a bit stressful to merge something too late. If @christoph-heiss uses nixos-unstable, maybe he doesnt care and prefers to wait

christoph-heiss committed:
nixos/sshd: add comment explaining different list option types

Signed-off-by: Christoph Heiss <christoph@c8h4.io>

christoph-heiss commented:
I would be fine with merging it now; it isn't a breaking change either.

@teto: I'm indeed on unstable, and the sooner this hits upstream the better, from my perspective.
